File: tests/tt_eager/ops/test_storage_type_validation.cpp
// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include <fmt/base.h>
#include <tt-metalium/constants.hpp>
#include <functional>
#include <gtest/gtest.h>

#include <tt-metalium/assert.hpp>
#include <tt-metalium/bfloat16.hpp>
#include <tt-metalium/device.hpp>
#include <tt-metalium/host_api.hpp>
#include <tt-metalium/shape.hpp>
#include "ttnn/decorators.hpp"
#include "ttnn/operations/eltwise/binary/binary.hpp"
#include "ttnn/operations/functions.hpp"
#include "ttnn/tensor/enum_types.hpp"
#include "ttnn/tensor/host_buffer/functions.hpp"
#include "ttnn/tensor/shape/shape.hpp"
#include "ttnn/tensor/storage.hpp"
#include "ttnn/tensor/tensor.hpp"
#include "ttnn/tensor/types.hpp"

using tt::tt_metal::DataType;
using tt::tt_metal::Layout;
using tt::tt_metal::Tensor;
using tt::tt_metal::distributed::MeshDevice;

class StorageTypeValidationTest : public ::testing::Test {
protected:
    void SetUp() override {
        int device_id = 0;
        device_owner = MeshDevice::create_unit_mesh(device_id);
        device = device_owner.get();
    }

    void TearDown() override { device_owner.reset(); }

    std::shared_ptr<MeshDevice> device_owner;
    MeshDevice* device{};
};

// Helper function to create host tensors and call binary operation
// This should trigger our storage type validation and throw a runtime error
void call_binary_with_host_tensors(bool use_legacy) {
    using tt::constants::TILE_HEIGHT;
    using tt::constants::TILE_WIDTH;

    ttnn::Shape shape({1, 1, TILE_HEIGHT, TILE_WIDTH});

    // Create host tensors using the proper method (like in copy_and_move test)
    auto random_data = ttnn::random::random(shape, DataType::BFLOAT16);
    auto host_tensor_a = random_data.cpu();  // Move to host
    auto host_tensor_b = random_data.cpu();  // Move to host

    // Verify these are host tensors
    TT_FATAL(
        host_tensor_a.storage_type() == tt::tt_metal::StorageType::HOST, "Input tensor A should be HOST storage type");
    TT_FATAL(
        host_tensor_b.storage_type() == tt::tt_metal::StorageType::HOST, "Input tensor B should be HOST storage type");

    // This should fail with storage type validation error and throw a runtime error
    auto result = ttnn::operations::binary::BinaryOperation<ttnn::operations::binary::BinaryOpType::ADD>::invoke(
        ttnn::DefaultQueueId,
        host_tensor_a,
        host_tensor_b,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        {},
        {},
        {},
        use_legacy);

    // If we get here, the validation failed (which is bad)
    TT_FATAL(false, "Expected storage type validation to fail, but operation succeeded");
}

TEST_F(StorageTypeValidationTest, BinaryOperationHostTensorValidation) {
    // Test that binary operation (use_legacy=true) throws runtime error when given host tensors
    // This is the expected behavior - our storage type validation should throw an exception
    EXPECT_THROW(call_binary_with_host_tensors(true), std::runtime_error);
}

TEST_F(StorageTypeValidationTest, BinaryNgOperationHostTensorValidation) {
    // Test that binary_ng operation (use_legacy=false) throws runtime error when given host tensors
    // This is the expected behavior - our storage type validation should throw an exception
    EXPECT_THROW(call_binary_with_host_tensors(false), std::runtime_error);
}

TEST_F(StorageTypeValidationTest, DeviceTensorsWorkCorrectly) {
    using tt::constants::TILE_HEIGHT;
    using tt::constants::TILE_WIDTH;

    ttnn::Shape shape({1, 1, TILE_HEIGHT, TILE_WIDTH});

    // Create tensors in TILE layout first, then move to device (should work)
    auto input_tensor_a = ttnn::random::random(shape, DataType::BFLOAT16).to_layout(Layout::TILE).to_device(device);
    auto input_tensor_b = ttnn::random::random(shape, DataType::BFLOAT16).to_layout(Layout::TILE).to_device(device);

    // Verify these are device tensors
    EXPECT_EQ(input_tensor_a.storage_type(), tt::tt_metal::StorageType::DEVICE);
    EXPECT_EQ(input_tensor_b.storage_type(), tt::tt_metal::StorageType::DEVICE);

    // This should work without crashing
    EXPECT_NO_THROW({
        auto result = ttnn::operations::binary::BinaryOperation<ttnn::operations::binary::BinaryOpType::ADD>::invoke(
            ttnn::DefaultQueueId,
            input_tensor_a,
            input_tensor_b,
            std::nullopt,
            std::nullopt,
            std::nullopt,
            {},
            {},
            {},
            true);
    });
}

TEST_F(StorageTypeValidationTest, BinaryNgOperationDeviceTensorsWorkCorrectly) {
    using tt::constants::TILE_HEIGHT;
    using tt::constants::TILE_WIDTH;

    ttnn::Shape shape({1, 1, TILE_HEIGHT, TILE_WIDTH});

    // Create tensors in TILE layout first, then move to device (should work)
    auto input_tensor_a = ttnn::random::random(shape, DataType::BFLOAT16).to_layout(Layout::TILE).to_device(device);
    auto input_tensor_b = ttnn::random::random(shape, DataType::BFLOAT16).to_layout(Layout::TILE).to_device(device);

    // Verify these are device tensors
    EXPECT_EQ(input_tensor_a.storage_type(), tt::tt_metal::StorageType::DEVICE);
    EXPECT_EQ(input_tensor_b.storage_type(), tt::tt_metal::StorageType::DEVICE);

    // This should work without crashing
    EXPECT_NO_THROW({
        auto result = ttnn::operations::binary::BinaryOperation<ttnn::operations::binary::BinaryOpType::ADD>::invoke(
            ttnn::DefaultQueueId,
            input_tensor_a,
            input_tensor_b,
            std::nullopt,
            std::nullopt,
            std::nullopt,
            {},
            {},
            {},
            false);
    });
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/ttnn/unit_tests/operations/eltwise/test_storage_type_validation.py
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
#
# SPDX-License-Identifier: Apache-2.0

import torch
import pytest
import ttnn
import random

# Set fixed random seeds for deterministic test execution
torch.manual_seed(42)


def test_binary_add_storage_type_validation_device_tensors(device):
    """Test that binary add (use_legacy=True) works correctly with device tensors."""
    # Create tensors on device (should work)
    x_torch = torch.tensor([[1.0, 2.0], [3.0, 4.0]], dtype=torch.float32)
    y_torch = torch.tensor([[0.5, 1.5], [2.5, 3.5]], dtype=torch.float32)

    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)

    # Verify these are device tensors
    assert x_tt.storage_type() == ttnn.StorageType.DEVICE
    assert y_tt.storage_type() == ttnn.StorageType.DEVICE

    # This should work without errors
    result = ttnn.add(x_tt, y_tt, use_legacy=True)
    result_torch = ttnn.to_torch(result)

    expected = x_torch + y_torch
    assert torch.allclose(result_torch, expected, atol=1e-5, rtol=1e-5)


def test_binary_ng_add_storage_type_validation_device_tensors(device):
    """Test that binary_ng add (use_legacy=False) works correctly with device tensors."""
    # Create tensors on device (should work)
    x_torch = torch.tensor([[1.0, 2.0], [3.0, 4.0]], dtype=torch.float32)
    y_torch = torch.tensor([[0.5, 1.5], [2.5, 3.5]], dtype=torch.float32)

    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)

    # Verify these are device tensors
    assert x_tt.storage_type() == ttnn.StorageType.DEVICE
    assert y_tt.storage_type() == ttnn.StorageType.DEVICE

    # This should work without errors
    result = ttnn.add(x_tt, y_tt, use_legacy=False)
    result_torch = ttnn.to_torch(result)

    expected = x_torch + y_torch
    assert torch.allclose(result_torch, expected, atol=1e-5, rtol=1e-5)


def test_binary_add_storage_type_validation_host_tensors():
    """Test that binary add (use_legacy=True) raises runtime error when given host tensors."""
    # Create tensors without device (should be host tensors)
    x_torch = torch.tensor([[1.0, 2.0], [3.0, 4.0]], dtype=torch.float32)
    y_torch = torch.tensor([[0.5, 1.5], [2.5, 3.5]], dtype=torch.float32)

    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT)

    # Verify these are host tensors
    assert x_tt.storage_type() == ttnn.StorageType.HOST
    assert y_tt.storage_type() == ttnn.StorageType.HOST

    # This should raise a runtime error due to our storage type validation
    with pytest.raises(RuntimeError, match="Input tensor A must be on device"):
        ttnn.add(x_tt, y_tt, use_legacy=True)


def test_binary_ng_add_storage_type_validation_host_tensors():
    """Test that binary_ng add (use_legacy=False) raises runtime error when given host tensors."""
    # Create tensors without device (should be host tensors)
    x_torch = torch.tensor([[1.0, 2.0], [3.0, 4.0]], dtype=torch.float32)
    y_torch = torch.tensor([[0.5, 1.5], [2.5, 3.5]], dtype=torch.float32)

    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT)

    # Verify these are host tensors
    assert x_tt.storage_type() == ttnn.StorageType.HOST
    assert y_tt.storage_type() == ttnn.StorageType.HOST

    # This should raise a runtime error due to our storage type validation
    with pytest.raises(RuntimeError, match="Input tensor A must be on device"):
        ttnn.add(x_tt, y_tt, use_legacy=False)
